### ISSUE TITLE
Segment Anything

### ISSUE THREAD

**Issue by nagataka** [nagataka/Read-a-Paper#55]
# Summary

#### Link

- [Segment Anything](https://arxiv.org/abs/2304.02643)
- https://github.com/facebookresearch/segment-anything

## What is this

- Image encoder: [ViTMAE](https://huggingface.co/docs/transformers/en/model_doc/vit_mae)
- Prompt encoder: CLIP
- Amodiﬁcation of a Transformer decoder block followed by a dynamic mask prediction head

## Key points

Data engine and its model-in-the-loop approach to obtain diverse and large segmentation dataset beyond existing ones (see "4. Segment Anything Data Engine")

1. Assisted-manual stage
    - In the first stage, resembling classic interactive segmentation, a team of professional annotators labeled masks by clicking foreground / background object points using a browser-based interactive segmentation
tool powered by SAM
        - collected 4.3M masks from 120k images in this stage
    - At the start of this stage, SAM was trained using common public segmentation datasets
2. Semi-automatic stage
    - In this stage, we aimed to increase the diversity of masks in order to improve our model’s ability to segment anything (focusing on annotating less prominent objects)
        - we first automatically detected confident masks (using an object detector; Faster R-CNN?). Then we presented annotators with images prefilled with these masks and asked them to annotate any additional
unannotated objects
3. Fully automatic stage## How the author proved effectiveness of the proposal?
    - 

## What should I read next?

[SAM2](https://github.com/facebookresearch/sam2)
[SAM3](https://github.com/facebookresearch/sam3)

## Code

Calculate the image embeddings using set_image(). Then this embedding is used 
https://github.com/facebookresearch/segment-anything/blob/main/segment_anything/predictor.py#L34

> By separating SAM into an image encoder and a fast prompt encoder / mask decoder, the same image embedding can be reused (and its cost amortized) with different prompts. Given an image embedding, the prompt encoder and mask decoder predict a mask from a prompt in ∼50ms in a web browser.

predict() takes point_coords, point_labels, or box as a prompt, it returns output masks, the model's predictions for the quality of each mask, and logits.

The reason output masks (plural) is potential ambiguity. Given a prompt it is possible that several different masks make sense. For this, SAM predicts multiple masks (the paper says 3 is sufficient in most cases.) "During training, we backprop only the minimum loss over masks. To rank masks, the model predicts a conﬁdence score (i.e., estimated IoU) for each mask."
